File: Python/ExampleTests/common/replace_post_string.py
from . import tokens
DNC_STRING = '?DNC?'


class ReplacePostString:
    def __init__(self, id: str = '', user_id: str = '', title: str = '', body: str = ''):
        self.id = id
        self.user_id = user_id
        self.title = title
        self.body = body

    @classmethod
    def from_list(cls, values):
        v = list(values)
        return cls(
            v[0] if len(v) > 0 else '',
            v[1] if len(v) > 1 else '',
            v[2] if len(v) > 2 else '',
            v[3] if len(v) > 3 else ''
        )

    @classmethod
    def from_text(cls, text):
        """Builds from the text form, e.g. Money as '25 USD'."""
        parts = tokens.require(text, 4, 'ReplacePost')
        return cls(
            parts[0],
            parts[1],
            parts[2],
            parts[3]
        )

    def __str__(self):
        return (tokens.token(self.id) + ' ' +
                tokens.token(self.user_id) + ' ' +
                tokens.token(self.title) + ' ' +
                tokens.token(self.body))

    def _key(self):
        return (self.id, self.user_id, self.title, self.body)

    def __eq__(self, other):
        if not isinstance(other, ReplacePostString):
            return NotImplemented
        return all(a == b or a == DNC_STRING or b == DNC_STRING
                   for a, b in zip(self._key(), other._key()))

    def __hash__(self):
        return hash(self._key())

    def __repr__(self):
        return f'ReplacePostString({self})'

File: Swift/ExampleTests/common/ReplacePostString.swift
public struct ReplacePostString: CustomStringConvertible, Equatable {
    public let id: String
    public let userId: String
    public let title: String
    public let body: String

    public init(id: String, userId: String, title: String, body: String) {
        self.id = id
        self.userId = userId
        self.title = title
        self.body = body
    }

    public init(fromArray v: [String]) {
        self.id = v.count > 0 ? v[0] : ""
        self.userId = v.count > 1 ? v[1] : ""
        self.title = v.count > 2 ? v[2] : ""
        self.body = v.count > 3 ? v[3] : ""
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    public static func fromText(_ text: String) -> ReplacePostString {
        let parts = Tokens.require(text, 4, "ReplacePost")
        return ReplacePostString(id: parts[0], userId: parts[1], title: parts[2], body: parts[3])
    }

    public var description: String {
        return Tokens.token(id) + " " + Tokens.token(userId) + " " + Tokens.token(title) + " " + Tokens.token(body)
    }

    public static let dncString = "?DNC?"

    public static func == (a: ReplacePostString, b: ReplacePostString) -> Bool {
        return (a.id == dncString || b.id == dncString || a.id == b.id)
            && (a.userId == dncString || b.userId == dncString || a.userId == b.userId)
            && (a.title == dncString || b.title == dncString || a.title == b.title)
            && (a.body == dncString || b.body == dncString || a.body == b.body)
    }
}
